config/src/main/java/com/alibaba/nacos/config/server/service/HistoryService.java

```diff
@@ -114,4 +119,60 @@ private void checkHistoryInfoPermission(ConfigHistoryInfo configHistoryInfo, Str
             throw new AccessException("Please check dataId, group or namespaceId.");
         }
     }
+
+    /**
+     * Query the detailed configuration history information pair, including the original version and the updated version.
+     */
+    public ConfigHistoryInfoPair getConfigHistoryInfoPair(String dataId, String group, String namespaceId, Long nid)
+            throws AccessException {
+        ConfigHistoryInfo configHistoryInfo = historyConfigInfoPersistService.detailConfigHistory(nid);
```

Comment:
should consider gray type history.

Reply:
OK

Reply:
> should consider gray type history.

check the history publish type，
if formal
-   optype-I： oldcontent - empty，newcontent- current history stored content
-   optype-U：oldcontent-current history stored content， newcontent- select next formal history content
-   optype-U:oldcontent -current history stored content,newcontent =empty
if gray
-   optype-I： oldcontent - empty，newcontent- current history stored content
-   optype-U：oldcontent-current history stored content， newcontent- select next gray history content with the same grayname
-   optype-U:oldcontent -current history stored content,newcontent =empty


replace the detailUpdatedConfigHistory method with. getNextHistoryInfo(String dataId,String group,String tenant, String publishType,String grayName,long startNid)

if grayName is empty
select * from his_config_info where dataid=? and group=? and tenant_id=? and publish_type=? and gray_name=?
and nid>startNid limit 1 

if grayName is not empty
select * from his_config_info where dataid=? and group=? and tenant_id=? and  publish_type=? and gray_name=?
and nid>startNid limit 1

Reply:
ok, i got it.